design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1072 @@
+
+## Design
+
+### `CREATE STEAM` and `CREATE TABLE`
+
+C* statements will, by default, capture the values of 
+`ksql.persistence.wrap.single.keys` and `ksql.persistence.wrap.single.values`.  
+These settings will be stored as part of the created source's metadata. 
+
+Users can override these settings by providing any of 
+the `WRAP_SINGLE_XXX` familty of `WITH` clause properties.
+
+These settings control both how the data in the source's topic should be
+deserialized if it only has a single field _and_ how downstream queries
+should serialize any single fields.  
+
+Note: as the settings control both deserialization and serialization, it is
+not possible to define a new source with a single field that should be 
+_deserialized_ unwrapped, but have downstream queries _serialize_ wrapped,
+or vice-versa. However, this limitation does allow for more simple syntax
+and downstream queries can override the setting as needed. (See rejected 
+alternatives for and alternative that uses two different settings for 
+deserialization and serialization).
+
+For example:
+
+```sql
+-- Override 'ksql.persistence.wrap.single.values' to true
+-- Indicating that the value of the data in the stream is wrapped. 
+-- Downstream queries will wrap single values when serializing them.  
+CREATE STREAM FOO (ID INT) WITH (WRAP_SINGLE_VALUES=true, ...); 
+
+-- Override 'ksql.persistence.wrap.single.values' to true
+-- As the value schema has multiple fields it will be deserialized wrapped. 
+-- Downstream queries will NOT wrap single values when serializing them.
+CREATE STREAM DAR (ID INT, NAME STRING) WITH (WRAP_SINGLE_VALUES=false, ...);
+
+-- [In future with structured keys]
+-- Override both settings to false.
+-- Indicating that the key and value data in the stream is not wrapped
+-- Downstream queries will not wrap single keys or values when serializing them.   
+CREATE TABLE BAR (ID INT KEY, NAME STRING) WITH (WRAP_SINGLE_FIELDS=false, ...); 
+``` 
+
+### `CREATE STREAM AS` and `CREATE TABLE AS`
+
+C*AS statements will inherit serialization settings from their source.
+In the case of joins, where there are multiple sources, the settings
+will be inherited from the left source. This follows the pattern used
+for other such `WITH` clause properties.
+
+For example:
+
+```sql
+-- Default config: ksql.persistence.wrap.single.values=false
+
+-- creates a stream, indicating that the source data has wrapped values
+-- and downstream queries will wrap values by default
+CREATE STREAM EXPLICIT_SOURCE (ID INT) WITH (WRAP_SINGLE_VALUES=true, ...);
+ 
+-- creates a stream, config indicating that the source data is not wrapped
+-- and downstream queries will NOT wrap values by default
+CREATE STREAM IMPLICIT_SOURCE (ID INT) WITH (...);
+
+-- creates a stream with multiple fields, so KSQL knows they'll be wrapped.
+-- and downstream queries will NOT wrap values by default
+CREATE STREAM MULTI_FIELD_SOURCE (ID INT, NAME STRING) WITH (WRAP_SINGLE_VALUES=false, ...);
```

Comment:
this one makes me nervous - it feels odd to dictate here, in the DDL, the output format of any subsequent query(ies) created against this source _if and only if those queries meet the condition that they have only a single field in the result value schema_.  As well as "feeling weird" (hardcore computer science term), it seems like a lot of corner-casing for the user to remember when trying to predict what will happen. 
i think i understand why you put this here, and the idea you are pushing on, but i'd propose that the `WRAP_SINGLE_VALUES...` stanza be inapplicable (rejected) for the DDL of a schema with multiple fields on the values side.

Reply:
I agree this is not a clear cut decision and is certainly one of the areas that needs some thought.

If we do not allow the user to set `WRAP_SINGLE_VALUES` here then we deny them the opportunity to say 

> look, I know this has multiple fields, but I'm going to build loads of queries off this source, and many will have single field schemas, and I want all of them to be serialized wrapped/unwrapped

I see no reason to allow this, hence including it.

Reply:
FYi, @hjafarpour raised a similar concern below: https://github.com/confluentinc/ksql/pull/2824#issuecomment-494019346